util/setuidgid.c

```diff
@@ -0,0 +1,329 @@
+			warnp("setgroups()");
+		goto err0;
+	}
+#endif
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Check if we're in any supplementary groups. */
+static int
+check_supplementary_groups_none()
+{
+	gid_t grouplist[1];
+	int ngroups;
+
+	/* Find number of groups.  Cannot fail with these arguments. */
+	ngroups = getgroups(0, NULL);
+
+	/* Check group membership. */
+	if (ngroups > 1)
+		/* Definitely failed. */
+		goto err0;
+	if (ngroups == 1) {
+		/*
+		 * POSIX allows getgroups() to return the effective group ID,
+		 * so if we're in exactly one group, we need to check that GID.
+		 */
+		if (getgroups(1, grouplist) != 1) {
+			warnp("getgroups()");
+			goto err0;
+		}
+		if (grouplist[0] != getegid())
+			goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Set the effective GID to the group represented by ${groupname}. */
+static int
+set_group(const char * groupname)
+{
+	gid_t gid;
+	struct group * group_info;
+
+	/* Find numerical GID. */
+	errno = 0;
+	if ((group_info = getgrnam(groupname)) == NULL) {
+		if (errno)
+			warnp("getgrnam(\"%s\")", groupname);
+		else
+			warn0("No such groupname: %s", groupname);
+		goto err0;
+	}
+	gid = group_info->gr_gid;
+
+	/* Set GID. */
+	if (setgid(gid)) {
+		/* We handle EPERM separately; keep it in errno. */
+		if (errno != EPERM)
+			warnp("setgid(%lu)", (unsigned long int)gid);
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Set the effective UID to the user represented by ${username}. */
+static int
+set_user(const char * username)
+{
+	uid_t uid;
+	struct passwd * user_info;
+
+	/* Calculate numerical UID. */
+	errno = 0;
+	if ((user_info = getpwnam(username)) == NULL) {
+		if (errno)
+			warnp("getpwnam(\"%s\")", username);
+		else
+			warn0("No such username: %s", username);
+		goto err0;
+	}
+	uid = user_info->pw_uid;
+
+	/* Set UID. */
+	if (setuid(uid)) {
+		/* We handle EPERM separately; keep it in errno. */
+		if (errno != EPERM)
+			warnp("setuid(%lu)", (unsigned long int)uid);
+		goto err0;
+	}
+
+	/* Success! */
+	return (0);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/* Parses "username", ":groupname", or "username:groupname". */
+static int
+string_extract_user_group(const char * combined, char ** username_p,
+    char ** groupname_p)
+{
+	size_t pos;
+	size_t len;
+
+	/* Sanity check. */
+	assert(combined != NULL);
+
+	/* Search for ":" and get string length. */
+	pos = strcspn(combined, ":");
+	len = strlen(combined);
+
+	/* Reject silly strings. */
+	if (pos == len - 1) {
+		warn0("Groupname is blank.");
+		goto err0;
+	}
+
+	/* String ok, proceed. */
+	if (pos == 0) {
+		/* Groupname only; duplicate string. */
+		if ((*groupname_p = strdup(&combined[1])) == NULL) {
+			warnp("strdup()");
+			goto err0;
+		}
+		*username_p = NULL;
+	} else if (pos != len) {
+		/* Extract username. */
+		if ((*username_p = malloc(pos + 1)) == NULL) {
```

Comment:
It occurred to me that instead of `malloc()`, I could use `strdup()` for everything:

1. make a copy of `const char * combined`
2. if the string looks good, replace `:` with `\0`
3. I can use `strdup()` for both the username and groupname

I think it would end up being more lines of code, so probably not worth it.  But I thought that would be a sneaky way to avoid `malloc`.  :)

Reply:
Yeah, "duplicate string, replace separators with NUL, duplicate individual substrings" is a common idiom -- especially in combination with `strtok`, which does the "replace separators with NUL" part.

I considered mentioning it but in this particular case I don't think it would make the code any clearer.